### PR TITLE
Implemented resumable downloads

### PR DESCRIPTION
**Pull request by noxer**
#### Summary
This PR adds download resume functionality to `mmetl`.
It verifies the download by adding a small overlap with the existing file and comparing the first 512 bytes.
The HTTP client now identifies itself via the User-Agent "mmetl/1.0".

```
[-----existing file-----]
                [-------resumed download-------]
                [overlap]
```

### CODE REVIEW

File: services/slack/intermediate.go [hunk 1]
```diff
@@ -493,7 +466,7 @@ func (t *Transformer) AddAttachmentsToPost(post *SlackPost, newPost *Intermediat
 	return props, propsByteArray
 }
 
-//func (t *Transformer) TransformPosts(slackExport *SlackExport, attachmentsDir string, skipAttachments, discardInvalidProps bool) error {
+// func (t *Transformer) TransformPosts(slackExport *SlackExport, attachmentsDir string, skipAttachments, discardInvalidProps bool) error {
```
Comment (agarciamontoro) [inline services/slack/intermediate.go:469]: What's this comment? :thinking:

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
This is neat, I really enjoyed reading your code :nerd_face:

I left some comments below, mostly nits, but also some about things I don't understand. I have a couple of general comments, too:
- Is it possible to test the functions in `download.go`? They seem brittle to any future change, so I'd love to have some assertions to ensure behavior is not changed accidentally in the future.
- Can we wrap the errors (when we just do `return err`) to give them a bit more context to what happened?

Thank you for your work on this, Tim! :)

File: services/slack/download.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	defer file.Close()
+
+	return resumeDownload(file, size, url)
+}
+
+func resumeDownload(existing *os.File, size int64, downloadURL string) error {
+	existingSize, overlap, err := calculateSize(existing, size)
+	if err != nil || existingSize == size {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (noxer) [reply services/slack/download.go:31]: I split it up and added a comment

File: services/slack/download.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+	start := existingSize - overlap
```
[the remark this answers is not in the document]
Comment (noxer) [reply services/slack/download.go:33]: Yeah, reading it now this is a bit confusing. Added a comment.

File: services/slack/download.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	if overlap != 0 {
+		err = checkOverlap(existing, resp.Body, overlap)
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err = existing.Seek(0, io.SeekEnd)
```
[the remark this answers is not in the document]
Comment (noxer) [reply services/slack/download.go:69]: We need to seek to the end. `checkOverlap` "consumes" the overlap and compares it. The new data after that is appended to the end.

File: services/slack/intermediate.go (hunk 1 above)
Comment (noxer) [reply services/slack/intermediate.go:469]: I have no idea. My `gofmt` changes this automatically. I think it should be removed completely.

File: services/slack/download.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (agarciamontoro) [reply services/slack/download.go:69]: Aaaaaah, I see. I thought `checkOverlap` only checked the overlap :nerd_face:

**Comment by noxer** [comment]
I'll look into the tests. It should be possible to "fake" the request. I'll also look into the error wrapping.

File: services/slack/download_test.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+	t.Run("successful resume from wrong file with an already downloaded file", func(t *testing.T) {
+		fileName := filepath.Join(os.TempDir(), "download-test")
+		defer os.Remove(fileName)
+		require.NoError(t, os.WriteFile(fileName, mockData, 0660))
+
+		require.NoError(t, downloadInto(fileName, srv.URL+"/wrong_resume", int64(len(mockData))))
+		tempFile, _ := os.ReadFile(fileName)
+		require.Equal(t, mockData, tempFile)
+	})
```
Comment (agarciamontoro) [inline services/slack/download_test.go:141]: Interesting case. Should we error there? There is no overlap between wrong file and already downloaded file. Can we make sure the one downloaded is the right one? In any case a super edge case, so up to you :)

**Review by agarciamontoro — APPROVED** [review]
Cool! Great tests, thank you :pray:

File: services/slack/download_test.go (hunk 5 above)
Comment (agarciamontoro) [reply services/slack/download_test.go:141]: Ah cool, then let's leave it as it is. Thanks for the context :)
